File: examples/src/mixing_simple.rs
//! # Simple Example: Mixing Simulation with `cmtool` Functionality
//!
//! **Purpose:**
//! This script demonstrates how to perform a simple mixing simulation in Rust,
//! analogous to the Python `pycmtool` library.
//! It integrates a mass balance over time using sparse matrix representations of the system,
//! with core functionality for chemical or compartmental mixing simulations.
//!
//! **Key Features:**
//! - Constructs flow sparse matrices from system states.
//! - Integrates mass balance equations
//!
//! **Example Workflow:**
//! 1. Define initial mass distribution.
//! 2. Advance the system state and construct sparse matrices.
//! 3. Integrate the system over a specified duration.
//!
//! Author:** CASALE Benjamin
//! Date:** 2025-09-18
//! Version:** 1.0

use cmtool_data::*;
use nalgebra::DMatrix;

// We use 2 species to demonstrates that pycmtool can handle different dissolved species
const N_SPECIES: usize = 2;

// fn get_transitionner<T: FlowMapTransitioner>() -> Option<T> {
//     let root = std::env::var("EXAMPLE_ROOT").unwrap();
//     let case_path = format!("{}/cma_case", root);
//     let p = std::path::Path::new(&case_path);
//     let case = CCMCaseInfo::read_case(p).unwrap();
//     //Load all the case information into the iterator
//     T::from_case(&root, &case).ok()
// }

fn integration<T: FlowMapTransitioner>(
    fm_t: &mut T,
    mass_0: &DMatrix<f64>,
    n_step: usize,
    duration: f64,
) -> DMatrix<f64> {
    let mut mass_i = mass_0.clone();
    let time_step = duration / ((n_step - 1) as f64);
    let mut current_time = 0.;
    for _ in 0..n_step {
        //Advance iterator to the corresponding flowmap (according to current_time or time_step)
        let it = fm_t.advance(current_time, time_step);
        //Get precalculated volume inverse
        let inverse_volume_j = &it.liquid.inverse_volume;
        //Convert sparse into dense matrix
        let transition = nalgebra::DMatrix::<f64>::from(it.liquid.get_transition());

        //Get concentration for mass and volume as C=M/V
        let concentration_i = {
            let mut tmp = mass_i.clone();
            (0..tmp.ncols()).for_each(|j| {
                tmp.column_mut(j).scale_mut(inverse_volume_j[j]);
            });
            tmp
        };
        //dm/dt = c*m
        mass_i += time_step * (concentration_i * transition);
        current_time += time_step;
    }

    mass_i
}

fn check_mixing<T: FlowMapTransitioner>(mut fm_t: T, final_time: f64, n_step: usize) -> Option<()> {
    let it = fm_t.get_current();

    let n_c = {
        let t = &it.liquid.transition;
        t.nrows()
    };

    //Get initial mass from volume and concentration
    let mass_0 = {
        let volume_0 = it.liquid.get_volume();
        let mut concentrations: nalgebra::DMatrix<f64> = nalgebra::DMatrix::zeros(N_SPECIES, n_c);
        concentrations[(0, 0)] = 1.;
        (0..concentrations.ncols()).for_each(|j| {
            concentrations.column_mut(j).scale_mut(volume_0[j]);
        });
        concentrations
    };

    let mass_i = integration(&mut fm_t, &mass_0, n_step, final_time);
    let mass_0m = mass_0.column_sum()[0];
    let mass_im = mass_i.column_sum()[0];

    let it = fm_t.get_current();
    let inverse_volume = &it.liquid.inverse_volume;

    let concentration_i = {
        let mut tmp = mass_i.clone();
        (0..tmp.ncols()).for_each(|j| {
            tmp.column_mut(j).scale_mut(inverse_volume[j]);
        });
        tmp
    };

    let mean_c = concentration_i.column_mean();
    let concentration_i_i_n = concentration_i.map(|x| x / mean_c[0]);

    println!("Inital mass: {}", mass_0m);
    println!("Final mass:  {}", mass_im);
    println!("Variance {}", concentration_i_i_n.column_variance()[0]);
    print!("Final normalized C: [");
    for i in 0..5 {
        print!("{}, ", concentration_i_i_n[(0, i)]);
    }
    println!("]");
    assert!((mass_0m - mass_im).abs() < 1e-8);
    Some(())
}

fn main() {
    let final_time: f64 = 50.;
    let n_step: usize = 5000;

    let root = std::env::var("EXAMPLE_ROOT").unwrap();
    //All fonctions use generic, specify iterator type here
    let t: DiscontinuousTransitioner = get_transitioner(&root).unwrap();
    check_mixing(t, final_time, n_step);
}

File: examples/python/mixing_simple.py
"""
Simple Example Script: Mixing Simulation with pycmtool
======================================================

**Purpose:**
This script demonstrates how to use the `pycmtool` library to perform a simple mixing simulation.
It integrates a mass balance over time using a sparse matrix representation of the system
The core functionality is used for chemical or compartmental mixing simulations.

**Key Features:**
- Uses `pycmtool` to construct flow sparse matrices from system states.
- Integrates mass balance equations using `scipy.integrate.solve_ivp`.
- Visualizes results with `matplotlib`.

**Example Workflow:**
1. Define initial mass distribution.
2. Use `pycmtool` to advance the system state and construct sparse matrices.
3. Integrate the system over a specified duration.
4. Plot the results.

**Author:** CASALE Benjamin
**Date:** 2025-09-18
**Version:** 1.0
"""

import os

import matplotlib.pyplot as plt
import numpy as np
import pycmtool
from scipy.integrate import solve_ivp

N_SPECIES = 2  # We use 2 species to demonstrates that pycmtool can handle different dissolved species


def integration(
    fmt: "pycmtool.Transitioner", mass_0: np.ndarray, duration: float, n_species
):
    """
    Perform a mixing simulation by integrating the mass balance ODE over a specified duration.

    This function uses `pycmtool` to advance the system state and construct sparse matrices,
    then integrates the mass balance equations using `scipy.integrate.solve_ivp`.

    Args:
        fmt: A `pycmtool.Transitioner` object to advance the system state.
        mass_0: Initial mass distribution (1D array of shape `(n_species * n_compartments,)`).
        duration: Total simulation time (in seconds or any consistent time unit).
        n_species: Number of species in the simulation .
    """

    def wrap(t: float, x: np.ndarray) -> np.ndarray:
        """Wrapper function for ODE integration."""
        d_t = 0  # Not used for this transitioner
        # Advance iterator to the corresponding flowmap (according to t or dt)
        it = fmt.advance(t, d_t)
        n_c = it.n_compartments
        liquid_state = it.liquid
        # Get the transition matrix
        transition = pycmtool.get_sparse_transition_matrix(liquid_state)

        vol = liquid_state.volumes  # Volume of each compartment
        _mass = x.reshape((n_species, n_c))  # Reshape to (N_SPECIES, n_compartments)
        C = _mass / vol  # Concentration: mass / volume
        return (C @ transition).reshape(-1)  # Return flattened array for ODE solver

    return solve_ivp(
        wrap,
        (0, duration),
        mass_0.reshape(-1),
        method="BDF",
        vectorized=False,
    )


def initial_c_distribution(n_c):
    """
    Generate a random initial concentration distribution for a simulation with `n_c` compartments.
    """
    return np.random.random((1, n_c))
    # m = np.zeros((n_c,))
    # m[0] = 1
    # return m


def get_normalized(it, y, i):
    """
    Calculate the normalized concentration for a given species across all compartments.

    The normalization is performed by dividing each compartment's concentration by the mean concentration
    of the species across all compartments. This is useful for comparing relative concentrations.
    """
    vol = it.liquid.volumes
    m0_c = y[0, :, i]
    return (m0_c / vol) / np.mean(m0_c / vol, axis=0)


def check_mixing(fmt, final_time: float):
    it = fmt.get_current()
    n_c = it.n_compartments

    C = np.zeros((N_SPECIES, n_c))
    C[0, :] = initial_c_distribution(n_c)
    vol = it.liquid.volumes
    m0 = C * vol

    sol = integration(fmt, m0, final_time, N_SPECIES)
    y = sol.y.reshape((N_SPECIES, n_c, -1))
    it = fmt.get_current()
    m0_c = y[0, :, 0]
    mt_c = y[0, :, -1]
    c_init = get_normalized(fmt.get_at(0), y, 0)
    c_final = get_normalized(fmt.get_at(fmt.n_flowmaps - 1), y, -1)

    m0m = np.sum(m0_c, axis=0)
    mfm = np.sum(mt_c, axis=0)

    print("Total volume: ", np.sum(vol))
    print("Inital mass: ", m0m)
    print("Final mass: ", mfm)
    print("Initial normalized C: ", c_init[:5])
    print("Final normalized C: ", c_final[:5])
    print("Final variance: ", np.var(c_final, axis=0))

    plt.figure()
    plt.style.use("tableau-colorblind10")
    plt.plot(c_final, label="final")
    plt.plot(c_init, "x", markersize=2.5, label="init")
    plt.title("Normalized concentration in all compartments (1 is the target value)")
    plt.xlabel("Compartment ID")
    plt.ylabel("Normalized concentration")
    plt.legend()

    plt.figure()
    plt.style.use("tableau-colorblind10")
    plt.plot(sol.t, y[0, 0, :])
    plt.legend()

    plt.show()
    assert abs(m0m - mfm) < 1e-8


if __name__ == "__main__":
    final_time = 200
    root = os.environ["EXAMPLE_ROOT"]
    # Let CMTool read and load the full case automatically, ready to iterate

    fmt = pycmtool.data.get_transitioner(root)
    check_mixing(fmt, final_time)
